### ISSUE TITLE
Potential Intermittent checks staying at queued

### ISSUE THREAD

**Issue by jusx** [jusx/mergeable#159]
When a **push** is made to a repos, GH will send a `check_suite.requested` event. In this event, the check status is set to `queued`

According to the [docs]( https://developer.github.com/v3/activity/events/types/#checksuiteevent):
> The `requested` action triggers when new code is pushed to the app's repository. 

For example, this event is received when a push is made to master and also when a push is made to the branch of a PR. 

We need to handle this especially so when this event is received and the payload has an array of pull_request that is not empty. This array contains all PRs that has the same head_branch and head_sha as the push (according to the docs). Not sure how it can ever be more than one element though.

### The problem
This may cause intermittent "hanging" checks in a PR — when a user pushes to the branch of the PR, two events are received by Mergeable: `pull_request.synchronize` and `check_suite.requested` — IF the `check_suite.requested` is received **after** mergeable finishes processing the pull_request.synchronize event and sent a status `completed` check, that check is potentially now in queue status.

There is a race-condition.

### Potential Solution
What we need to do when check_suite.requested event is received:

- Add this event to special cases function and get the pull_request. For now only get the first element and assume there's ever only one element in the pull_request array.
- Add check_suite.requested to the checks action.
- If there are config errors call Checks.run
- else processWorkflow.

## Update 2019-02-07
Did a little bit more testing with various scenarios:

1. As long as checks are created and updated with complete or cancelled status, everything looks normal from a user perspective and there is no need to take action on the check_suite.requested event for that PR. It works as expected.

2. If there are no checks manually created and completed for a PR where the check_suite.requested event is triggered then the UI will not show anything in the conversations tab, however it will show a queued status for the app in the checks tab. It does not prevent merging though.

Currently, mergeable will always create a check and finish it with completed status for every event it is asked to listen to in the configuration -- regardless of whether the check_suitre.requested is received. This is the expected behavior.

**There isn't any race conditions** that I can determine since we do `1`. The check_suite.requested event has a totally different check run id as compared to the one we receive when creating a check_suite.

**Comment by denis-sokolov** [comment]
I suspect this is also the cause of my experience of forever queued checks:

![image](https://user-images.githubusercontent.com/113721/52358760-43c90880-2a41-11e9-9301-384671a45508.png)

When I open a PR, the Mergeable check runs successfully and becomes green. When a reviewer approves it, the check turns yellow. This is the relevant piece from my Mergeable configuration:

```yaml
version: 2
mergeable:
  - when: pull_request.opened, pull_request.edited, pull_request.synchronize
    validate:
```

**Comment by ndench** [comment]
I was talking to @jusx on [gitter](https://gitter.im/mergeable-bot/Lobby) about this issue. He mentioned I should try this:
```
- when: pull_request.*, pull_request_review.*
```
Which has made the problem much less prevalent, to the point where we never notice it anymore (because everytime anything happens it just reruns mergeable)

**Comment by jusx** [comment]
@denis-sokolov @ndench Thanks for the comments. Based on the reports from the guys at  [Read the Docs](https://github.com/rtfd) and looking at [their PRs](https://github.com/rtfd/readthedocs.org/pulls) with pending checks I was able to narrow the problem down deeper into this https://github.com/jusx/mergeable/issues/168

It looks like pretty much all of the problem is coming from that. 

However, there is still a chance for this race condition. Just much smaller.

**Comment by jusx** [comment]
I've updated the description with an update and more findings. Sine the checksuite APIs are developer preview and as noted in their docs could change, I've contacted GH to confirm that:

- As long as mergeable creates a check and updates it with a complete status, the auto triggered check_suite.requested event can be ignored. 

The current behavior shows that this is true but I want to confirm this will not change in the future.

**Comment by jusx** [comment]
Also, all of the forever queued check problems with mergeable should have been resolved with the fix for #168

**Comment by jusx** [comment]
I heard back from the folks at GitHub that the auto triggers with `* check_suite.requested` can be ignored as long as a separate check is created/updated with "completed" status. There is no race condition here. 

Closing this issue.

@denis-sokolov I believe your problem is resolved with #168 -- let me know if you're still seeing the problem. Thanks!

**Comment by mshwery** [comment]
I'm actually witnessing this same thing still but am using `- when: pull_request.*, pull_request_review.*` already 🤔
